### ISSUE TITLE
Signalr Extensions using Class-based model is not currently supported

### ISSUE THREAD

**Issue by michaeldaw** [Azure/azure-functions-dotnet-worker#771]
Worker.Extensions.SignalRService does not currently contains classes required to develop Functions using the [Class based model](https://docs.microsoft.com/en-us/azure/azure-signalr/signalr-concept-serverless-development-config#class-based-model).
eg:
 - ServerlessHub
 - InvocationContext
 - SignalRConnectionInfo
  
Are there plans to support the Class Based Model when using Worker.Extensions.SignalRService?

**Comment by fabiocav** [comment]
@Y-Sindo have you had a chance to look at this? Some of them should be very simple to bring in.

**Comment by Y-Sindo** [comment]
I will add a PR to provide some convenient classes such as `SignalRConnectionInfo`, `SignalRInvocationContext` soon. 

About the serverless hub in isolated-process model, we do have a plan to support it, but we still have some problems to overcome.

**Comment by Y-Sindo** [comment]
@fabiocav I have added a PR to address the issue, please take a look at it.

**Comment by fabiocav** [comment]
Fantastic! Thanks @Y-Sindo. We'll review that ASAP.

FYI @anthonychu

**Comment by michaeldaw** [comment]
Sorry to interrupt the discussion. Just wanted to say thanks for the response.

**Comment by Arash-Sabet** [comment]
@fabiocav @anthonychu Any idea when the PR associated with this issue will be approved and released? Thanks.

**Comment by Y-Sindo** [comment]
@fabiocav @kshyju Currently the SignalR class-based model can bind the RPC call parameters to trigger function parameters automatically, I want to know if the bindings in host process can get the parameter info of the trigger functions in worker process, so that we can do similar things in isolated model.

**Comment by liliankasem** [comment]
@kshyju looks like the PR was merged last week, any idea on a potential release date?

**Comment by michaeldaw** [comment]
Thanks to everyone who's worked on this. Will it be part of a release soon? I've started a project that could really use this 😊

**Comment by duncanthescot** [comment]
What is the current status of this?  The last message I see is from a year ago.

**Comment by Y-Sindo** [comment]
Please see this project for an example to integrate with SignalR management SDK https://github.com/aspnet/AzureSignalR-samples/tree/main/samples/DotnetIsolated-ClassBased

**Comment by Y-Sindo** [comment]
@fabiocav Seems that the host process cannot get the real parameters of functions in isolated worker, then I'd like to know if there is a way to bind a value to the parameters not annotated with any attribute. For example, in the SignalR extension of in-process model, as following code sample, the `parameter1` `parameter2` are the arguments of a SignalR client remote invocation, and the values of them are also provided by the SignalR trigger provider. However, during my debug process, I find that if this function is defined in isolated worker, then the host process can only see the first argument `invocationContext`.
```cs
        [Function("Broadcast")]
        public Task Broadcast(
        [SignalRTrigger] SignalRInvocationContext invocationContext, string parameter1, string parameter2 )
        {
        }
```

**Comment by michaeldaw** [comment]
@Y-Sindo If we're still talking about non-class-based mode, I've had to declare the function trigger as in the example below in order to ensure the parameters are correctly bound:
```cs
    [Function("MessageFromClient")]
    public async Task ClientCommand(
        [SignalRTrigger("hub", "messages", "MessageFromClient", "parameter1", "parameter2")]
        SignalRInvocationContext invocationContext, string parameter1, string parameter2)
    {
    }
```
To be clear, this is for a trigger in which a message is coming from the client to the server. The parameters of the `SignalRTrigger` attribute are "hubName", "category", "eventName", followed by a params list of parameter names.

**Comment by michaeldaw** [comment]
In any case, since creating this issue I've learned to live with the non-class-based mode 😅.

**Comment by Y-Sindo** [comment]
@michaeldaw 

> @Y-Sindo If we're still talking about non-class-based mode, I've had to declare the function trigger as in the example below in order to ensure the parameters are correctly bound:
> 
> ```cs
>     [Function("MessageFromClient")]
>     public async Task ClientCommand(
>         [SignalRTrigger("hub", "messages", "MessageFromClient", "parameter1", "parameter2")]
>         SignalRInvocationContext invocationContext, string parameter1, string parameter2)
>     {
>     }
> ```
> 
> To be clear, this is for a trigger in which a message is coming from the client to the server. The parameters of the `SignalRTrigger` attribute are "hubName", "category", "eventName", followed by a params list of parameter names.

Yes, you're right. To remove the requirement of declaring the method name and parameter names is one of the tasks in serverless hub. Currently I've encountered an issue when trying to implement it. See my earlier comment for detail: https://github.com/Azure/azure-functions-dotnet-worker/issues/771#issuecomment-1524641046

 > Edit: One thing I have not been able to do successfully is return a value to the client from a trigger like the above. In earlier projects (before I switched to Isolated-mode), I was able to specify a return value for the function. The return value would be serialized and sent back to the client. Since using isolated mode, I've found that returns values are null/undefined when they get to the client.

Confirmed the problem. I have created a new issue #1496 to check the problem. It still needs more investigation.

Thank you for sharing your experience. We appreciate your adaptability in working with the non-class-based mode. If you have any further issues or suggestions, please feel free to let us know. We value your feedback as we continue to improve our software.
